conduit/metrics.py

```diff
@@ -0,0 +1,202 @@
+from enum import Enum
+from functools import partial
+from typing import Generic, List, Optional, Protocol, Tuple, TypeVar, Union
+
+from ranzen import str_to_enum
+import torch
+from torch import Tensor
+
+__all__ = [
+    "Comparator",
+    "accuracy",
+    "accuracy_per_subclass",
+    "hard_prediction",
+    "per_subclass_metric",
+    "precision_at_k",
+    "robust_accuracy",
+    "robust_gap",
+    "subclass_balanced_accuracy",
+    "tnr_per_subclass",
+    "tpr_per_subclass",
+    "tpr_differences",
+]
+
+
+@torch.no_grad()
+def hard_prediction(logits: Tensor) -> Tensor:
+    logits = torch.atleast_2d(logits.squeeze())
+    return (logits > 0).long() if logits.ndim == 1 else logits.argmax(dim=1)
+
+
+@torch.no_grad()
+def accuracy(y_pred: Tensor, *, y_true: Tensor) -> Tensor:
+    y_pred = torch.atleast_2d(y_pred.squeeze())
+    if len(y_pred) != len(y_true):
+        raise ValueError("'logits' and 'targets' must match in size at dimension 0.")
+    preds = hard_prediction(y_pred)
+    return (preds == y_true).float().mean()
+
+
+@torch.no_grad()
+def precision_at_k(
+    y_pred: Tensor, *, y_true: Tensor, top_k: Union[int, Tuple[int, ...]] = (1,)
+) -> List[Tensor]:
+    """Computes the accuracy over the k top predictions for the specified values of k"""
+    if isinstance(top_k, int):
+        top_k = (top_k,)
+    maxk = max(top_k)
+    batch_size = y_true.size(0)
+    _, pred = y_pred.topk(k=maxk, dim=1, largest=True, sorted=True)
+    pred = pred.t()
+    correct = pred.eq(y_true.view(1, -1).expand_as(pred))
+
+    res: List[Tensor] = []
+    for k in top_k:
+        correct_k = correct[:k].contiguous().view(-1).float().sum(0, keepdim=True)
+        res.append(correct_k.mul_(100.0 / batch_size))
+    return res
+
+
+class Comparator(Protocol):
+    def __call__(
+        self, y_pred: Tensor, *, y_true: Tensor
+    ) -> Tuple[Tensor, Optional[Union[Tensor, slice]]]:
+        """
+        :param y_pred: Predicted labels or raw logits of a classifier.
+        :param y_true: Ground truth (correct) labels.
+
+        :returns: An element-wise comparison between ``y_pred`` and ``y_true`` or a subset of them;
+        if the latter, the second element returned should be a mask indicating which samples
+        comprise that subset.
+        """
+        ...
+
+
+def _pdist_1d(x: Tensor) -> Tensor:
+    return torch.pdist(x.view(-1, 1)).squeeze()
+
+
+def _max_diff_1d(x: Tensor) -> Tensor:
+    return _pdist_1d(x)
+
+
+class Aggregator(Enum):
+    MIN = min
+    "Aggregate by taking the minimum."
+    MAX = max
+    "Aggregate by taking the maximum."
+    MEAN = torch.mean
+    "Aggregate by taking the mean."
+    MEADIAN = torch.median
+    "Aggregate by taking the median."
+    DIFF = partial(_pdist_1d)
+    "Aggregate by taking the pairwise (absolute) differences."
+    MAX_DIFF = partial(_max_diff_1d)
```

Comment:
Why are these partials? Is it because otherwise the functions become methods or something like that?

Reply:
I have no idea why the partials aren't required for my own functions but not for the built-in and torch ones, but alas that's how it is (specifically without partial application the values seemingly become the names, bafflingly)

Reply:
You could use the trick I used in EthicML which is to use a 1-tuple to wrap the callables.

Reply:
That works?

Reply:
Ah but you do need to unpack if you do it that way

Reply:
with partial, on the other hand, I can simply call the value which means there's no need for special-casing

Reply:
Well, you would wrap all the enum values, so when you call the functions, you'd do `e.value[0](args)`. But I can understand if you prefer the `partial` solution

Reply:
It's a general problem with assigning functions to variables in class bodies, not just Enum: https://stackoverflow.com/a/28415970